runtime/cudaq/platform/default/rest_server/RemoteSimulatorQPU.cpp

```diff
@@ -0,0 +1,357 @@
+        }
+      }
+    } else {
+      // Get the quake representation of the kernel
+      auto quakeCode = cudaq::get_quake_by_name(name);
+      request.passes = serverPasses;
+      request.format = cudaq::CodeFormat::MLIR;
+      auto module = parseSourceString<ModuleOp>(quakeCode, &m_mlirContext);
+      if (!module)
+        throw std::runtime_error("module cannot be parsed");
+
+      // Extract the kernel name
+      auto func = module->lookupSymbol<mlir::func::FuncOp>(
+          std::string("__nvqpp__mlirgen__") + name);
+
+      // Create a new Module to clone the function into
+      auto location = FileLineColLoc::get(&m_mlirContext, "<builder>", 1, 1);
+      ImplicitLocOpBuilder builder(location, &m_mlirContext);
+
+      // FIXME this should be added to the builder.
+      if (!func->hasAttr(cudaq::entryPointAttrName))
+        func->setAttr(cudaq::entryPointAttrName, builder.getUnitAttr());
+      auto moduleOp = builder.create<ModuleOp>();
+      moduleOp.push_back(func.clone());
+
+      if (args) {
+        cudaq::info("Run Quake Synth.\n");
+        PassManager pm(&m_mlirContext);
+        pm.addPass(cudaq::opt::createQuakeSynthesizer(name, args));
+        if (failed(pm.run(moduleOp)))
+          throw std::runtime_error("Could not successfully apply quake-synth.");
+      }
+
+      // Client-side passes
+      if (!clientPasses.empty()) {
+        PassManager pm(&m_mlirContext);
+        std::string errMsg;
+        llvm::raw_string_ostream os(errMsg);
+        const std::string pipeline =
+            std::accumulate(clientPasses.begin(), clientPasses.end(),
+                            std::string(), [](const auto &ss, const auto &s) {
+                              return ss.empty() ? s : ss + "," + s;
+                            });
+        if (failed(parsePassPipeline(pipeline, pm, os)))
+          throw std::runtime_error(
+              "Remote rest platform failed to add passes to pipeline (" +
+              errMsg + ").");
+
+        if (failed(pm.run(moduleOp)))
+          throw std::runtime_error(
+              "Remote rest platform: applying IR passes failed.");
+      }
+      std::string mlirCode;
+      llvm::raw_string_ostream outStr(mlirCode);
+      mlir::OpPrintingFlags opf;
+      opf.enableDebugInfo(/*enable=*/true,
+                          /*pretty=*/false);
+      moduleOp.print(outStr, opf);
+      request.code = llvm::encodeBase64(mlirCode);
+    }
+
+    request.simulator = m_simName;
+    request.seed = cudaq::get_random_seed();
+    // Don't let curl adding "Expect: 100-continue" header, which is not
+    // suitable for large requests, e.g., bitcode in the JSON request.
+    //  Ref: https://gms.tf/when-curl-sends-100-continue.html
+    std::map<std::string, std::string> headers{
+        {"Expect:", ""}, {"Content-type", "application/json"}};
+    json requestJson = request;
+    auto resultJs = m_client.post(m_url, "job", requestJson, headers, false);
```

Comment:
How should we handle exceptions thrown in the server? Maybe we can grab the message there and that can be part of the messaged received here by the client.

Reply:
This has been implemented. The server will now report a 500 error code to the client which then throws an exception.